### PR TITLE
feature: personal file view

### LINKED ISSUE #42919: 🗃️ "Personal files" view
## Context
With heavy use of Group Folders, the `All Files` menu view is confusing for the user who gets his/her personal files mixing with organization files. Currently there is no option on top level to have filtered view only showing a user their own files.

### Describe the solution you'd like
Add an entry `Personal files` to the left menu, directly below `All files`, showing:
- only user files
- no groupfolders
- no items shared from other people

### PR DESCRIPTION
**Pull request by emoral435**
## Issue Summary
Resolves #42919 

#### Quick GIF View
firefox_UlvxgyFkts
#### File sharing comparison
🏚️ Admin View | 🏡 Janes View
---|---
admin all files| jane all files
image | image
admin personal files (root) | jane personal files (root
image | firefox_KumFQsCbr2

### Clarifications
- User clicks on "Personal Files" from the Sidebar, and using folder API's, filters group-folders, shared folders/files that do not belong to the User
- When a folder is shared to another person, any file or folder that person makes belongs to whoever the 'root' folder belongs too

### Implementation
- Added new view to the files view that filters each file/folder that the user uploaded / created / owns
- Each view from there filters out each sub-file that they did not upload / create / own, but only is really important at the root level

### TODO
- [ ] Add tests to personal views

### Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required (NOT REQUIRED)
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by emoral435** [comment]
Using this just to draft my PR for the personal files view :) no reviews needed yet!

**Comment by emoral435** [comment]
### TLDR for the reviewers: 

- Need a review from the design team to know I can go forward with making tests for the view 👍
- Checkup with @tobiasKaminsky to see if our assumptions are similar so that the mobile and web team can be on the same page
- My fellow frontend reviewers to absolutely cook this PR for any flawed logic 🙏🙈

If anything needs clarification, just let me know!

**Comment by nimishavijay** [comment]
Nice! Looking good! :) I would suggest that a heading be added to the personal files view, otherwise I imagine users can get lost, or are confused about why a file disappeared suddenly. 

It can be a part of the home icon in the breadcrumbs bar. So would it be possible to make it look like a icon+text tertiary button, like `[home icon] Personal files`? :)

**Review by jancborchardt — DISMISSED** [review]
Really nice! :)

Small detail, as @nimishavijay already mentioned the capitalization needs to be "Personal files" ("files" in lowercase).

And the icon could be a bit simplified to https://pictogrammers.com/library/mdi/icon/account/

**Comment by emoral435** [comment]
Both of these changes are wonderful details to either change / improve on! I can definitely add both suggestions.

I agree that differentiating the views by breadcrumb navigation is something that would be a small change, but a good stylistic one at that. Will adjust the PR accordingly.

**Comment by emoral435** [comment]
firefox_QiHyTJ2z9W

@nimishavijay This is how the implementation looks! IMO, I think that having this now makes this view look a bit different from the rest - I think that having the rest of the views have something similar where we append the view name at the end of the file name would now make this much cleaner, no? This is certainly possible as well.

File: apps/files/src/views/personal-files.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import { translate as t } from '@nextcloud/l10n'
+import { View, getNavigation } from '@nextcloud/files'
+
+import { getContents } from '../services/PersonalFiles'
+import AccountIcon from '@mdi/svg/svg/account.svg?raw'
+import logger from '../logger'
+
+export default () => {
+	logger.debug("Loading root level personal files view...")
+
```
Comment (skjnldsv) [inline apps/files/src/views/personal-files.ts:31]:
```suggestion
```

File: apps/files/src/services/PersonalFiles.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+					} catch (error) {
+						logger.error(`Invalid node detected '${result.basename}'`, { error })
+						return null
+					}
+				}).filter(node => davNotShared(node, currUserID)) as File[],
+			})
+        } catch (error) {
+            reject(error)
+        }
+    })
```
Comment (skjnldsv) [inline apps/files/src/services/PersonalFiles.ts:87]: Let's directly import `getContents` from the Files service and chain the promise to filter the results.

File: apps/files/src/services/PersonalFiles.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ * NOTE MOVE TO @nextcloud/files 
+ * @brief filters each file/folder on its shared statuses
+ * @param {FileStat} node that contains  
+ * @return {Boolean}
+ */
+export const davNotShared = function(node: File | Folder | null, currUserID: string | undefined): Boolean {
+	// (essentially .filter(Boolean))
+	if (!node) return false
+	
+	const isNotShared = currUserID ? node.attributes['owner-id'] === currUserID : true
```
Comment (skjnldsv) [inline apps/files/src/services/PersonalFiles.ts:44]: I would use `node.owner`

Comment (skjnldsv) [reply apps/files/src/services/PersonalFiles.ts:44]: We fixed the owner not long ago https://github.com/nextcloud/server/commit/84813935812469994cfacfdcb876946891117114#diff-ecc6561845b0675ca8ca40c4ee0f161d3b2ec010b6bfcfdafa28123fa5273d13

**Comment by emoral435** [comment]
Utilized the getContents from the Files view, then filtered from there. Simplified the codebase a lot 🐘

**Review by Pytal — APPROVED** [review]
Code looks good!

File: apps/files/src/services/PersonalFiles.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * 		b.) the file is not shared with anyone
+ * 		c.) the file is not a group folder
+ * @param {FileStat} node that contains  
+ * @return {Boolean}
+ */
+export const personalFile = function(node: File): File | null {
+	const isNotShared = currUserID ? node.owner === currUserID : true
+						&& node.attributes['mount-type'] !== 'group'
+						&& node.attributes['mount-type'] !== 'shared'
+	return isNotShared ? node : null
```
Comment (Pytal) [inline apps/files/src/services/PersonalFiles.ts:43]:
```suggestion
	return isNotShared
```

File: apps/files/src/services/PersonalFiles.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+import { File, type ContentsWithRoot } from '@nextcloud/files'
+import { getCurrentUser } from '@nextcloud/auth';
+
+import { getContents as getFiles } from './Files';
+
+const currUserID = getCurrentUser()?.uid.toString()
```
[the remark this answers is not in the document]
Comment (emoral435) [reply apps/files/src/services/PersonalFiles.ts:27]: Sorry for barely getting to this now, was sick last Friday, but resolved!

File: apps/files/src/services/PersonalFiles.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+export const getContents = (path: string = "/"): Promise<ContentsWithRoot> => {
+	// get all the files from the current path as a cancellable promise
+	// then filter the files that the user does not own, or has shared / is a group folder
+    return getFiles(path)
+			.then(c => {
+				c.contents = c.contents.filter(personalFile) as File[]
+				return c
+			})
+}
```
Comment (skjnldsv) [inline apps/files/src/services/PersonalFiles.ts:54]:
```suggestion
	return getFiles(path)
		.then(c => {
			c.contents = c.contents.filter(personalFile) as File[]
			return c
		})
}

```

**Review by skjnldsv — APPROVED** [review]
One lint issue, otherwise LGTM

**Review by AndyScherzinger — APPROVED** [review]

**Comment by emoral435** [comment]
Enabling auto-merge to hopefully get the feature out by Thursday night 👍

File: apps/files/src/services/PersonalFiles.ts (hunk 6 above)
Comment (emoral435) [reply apps/files/src/services/PersonalFiles.ts]: resolved!
